src/OpenTelemetry.Api/CHANGELOG.md

```diff
@@ -2,6 +2,16 @@
 
 ## Unreleased
 
+* **Breaking change:** `Baggage.Current` behavior changed to match `AsyncLocal`
+  (copy-on-write). Changes now create a new
+  [ExecutionContext](https://learn.microsoft.com/dotnet/api/system.threading.executioncontext)
+  and will only be available downstream. Previously changes also flowed back to
+  parent contexts.
+  ([#5208](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5208))
+
+* Added `Baggage.TryGetBaggage` API.
```

Comment:
Does this need to be public?

Reply:
It is just a convenience thing. BUT I would say it is also a standard thing in the .NET world one would expect on dictionary-like classes a la [TryGetValue](https://learn.microsoft.com/dotnet/api/system.collections.generic.dictionary-2.trygetvalue).

```csharp
if (!Baggage.TryGetBaggage("TenantId", out var tenantId))
{
    tenantId = "Unknown";
}

DoSomethingWithTenantId(tenantId);
```

vs

```csharp
string? tenantId = Baggage.GetBaggage("TenantId"); // May return null if not found. Users need to code properly to handle that case
if (tenantId == null)
{
    tenantId = "Unknown";
}

DoSomethingWithTenantId(tenantId);
```

> when should the user choose GetBaggage vs. TryGetBaggage

They should always choose `TryGetBaggage` because it is less error prone but up until now we didn't offer it!

Reply:
> It is just a convenience thing. BUT I would say it is also a standard thing in the .NET world one would expect on dictionary-like classes a la [TryGetValue](https://learn.microsoft.com/dotnet/api/system.collections.generic.dictionary-2.trygetvalue).

This is unclear to me. Dictionary has `TryGetValue` because `operator []` would throw if the key is not found. For Baggage, null will be returned for any key that doesn't exist. Is this aligned with existing practice?

Reply:
IMO `GetBaggage` is error prone because it breaks with `Dictionary` (doesn't throw) and expects users to read the XML docs to understand that a `null` might be spit back. The nullable annotations being added here should help point out mistakes for anyone targeting >= .NET Core 3.0 but I do feel `TryGetBaggage` is better (less error prone, more obvious behaviors).

If you still aren't convinced LMK and I'll remove it.

Reply:
> ... because it breaks with `Dictionary` ...

Would you elaborate on why this is considered as breaking?

I don't think anyone should consider Baggage as a dictionary, the right expectation is captured by https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/baggage/api.md#get-value: _"... the Baggage API MUST provide a function that takes the name as input, and returns a value associated with the given name, or null if the given name is not present ..."_

Reply:
Do we expect our users to be familiar with the spec? I suspect most have never seen it 🤣

We should follow the spec, of course, but I think slight deviations in order to present an idiomatic API surface area for .NET are well within the purview of the SIG.

Reply:
I disagree, users who are using Baggage APIs are expected to understand what Baggage is - it's part of their job. We don't want users to "just use Baggage as dictionary without understanding the concept". They might read some articles instead of the spec though.

Reply:
Updated to remove `TryGetBaggage`. Wholesale disagree with your thinking though 🤣 Safe-by-default API is always > API which requires spec reading to use correctly ⛑️